docs: add bun runtime explanations (#7054)
dagger/dagger: docs/current_docs/manuals/developer/typescript/356352-runtime.mdx

---
slug: /manuals/developer/typescript/356352/runtime
displayed_sidebar: "current"
toc_max_heading_level: 2
title: "Runtimes"
---

# Runtimes

> Multiple runtimes support TypeScript such as [Node.js](https://nodejs.org/en), [Deno](https://deno.com/) and [Bun](https://bun.sh/).

Reviewer: I think we should correct this. @TomChv?

> Multiple runtimes support TypeScript such as [Node.js](https://nodejs.org/en), [Deno](https://deno.com/) and [Bun](https://bun.sh/).

Author: `tsx` is based on Node.JS runtime no?

> Multiple runtimes support TypeScript such as [Node.js](https://nodejs.org/en), [Deno](https://deno.com/) and [Bun](https://bun.sh/).

Reviewer 2: @TomChv 
Yes it is. I think the current statement conveys the right message, even though Node does not support typescript. It is just that we use tsx. tsx just handles all the ts compalition(I think using esbuild) and then runs the code using node. Might just be something we would like to add here.

> Multiple runtimes support TypeScript such as [Node.js](https://nodejs.org/en), [Deno](https://deno.com/) and [Bun](https://bun.sh/).

Reviewer: Maybe reword it to 

TypeScript is supported by multiple runtimes: [Node.js](https://nodejs.org/en) (via `tsx`), [Deno](https://deno.com/) (natively) and [Bun](https://bun.sh/) (natively).

> Multiple runtimes support TypeScript such as [Node.js](https://nodejs.org/en), [Deno](https://deno.com/) and [Bun](https://bun.sh/).

Reviewer: Added as suggestion

By default, the TypeScript SDK executes functions using the Node.js runtime, but you can configure an alternative runtime in the `package.json` file.

:::note
For now, the TypeScript SDK only supports Node.js (stable) and Bun (experimental).
:::

To change the TypeScript SDK runtime, set the field `dagger.runtime` in the `package.json` file.

## Node  

Set the field to `node` to use the Node.js runtime. Node.js is also the default runtime used if this field is omitted.

```json
  "dagger": {
    "runtime": "node"
  }
```

## Bun 

Set the field to `bun` to use the Bun runtime.

```json
  "dagger": {
    "runtime": "bun"
  }
```

:::warning
Bun runtime support is still experimental. Unexpected results may occur in some cases.

> :::

Reviewer 2: ```suggestion
:::
## Automatic Detection
When a runtime is not explicitly defined within the `package.json` file, Dagger automatically identifies the appropriate runtime by examining the project's lock files inside the projects `dagger` directory.

- If Dagger finds any of the following lock files : `package-lock.json`, `yarn.lock`, or `pnpm-lock.yaml`, it automatically selects `node` as the runtime.

- In the absence of the lock files mentioned above, if a `bun.lockb` file is present, Dagger will choose `bun` as the runtime.

- If no or only unknown lock files are present `node` is choosen.

This behavior however should be seen as a sensible fallback and not as an explicit configuration.
```

We could add something like this at the end. This way potential confusions could be avoided. F.e someone uses bun as a package manager, but still wants to use the Node.js runtime. This section explains daggers behavior in this situation.
I am also open to not include this for the sake of simplicity.

Reviewer: This seems like valuable information to add.

---
slug: /manuals/developer/typescript/356352/runtime
displayed_sidebar: "current"
toc_max_heading_level: 2
title: "Runtime (Experimental)"
---

# Runtime

Multiple runtimes support TypeScript such as [NodeJS](https://nodejs.org/en), [Deno](https://deno.com/) and [Bun](https://bun.sh/).

By default, the TypeScript SDK executes functions using the NodeJS runtime but you can configure a different runtime in the `package.json`.

:::note
For now, the TypeScript SDK only supports NodeJS and Bun.
:::

To change the Typescript SDK runtime, set the field `dagger.runtime` in the `package.json`.

## Node  

Set the field to `node` to use the `NodeJS` runtime, it's also the default runtime used if this field is ommited.

```json
  "dagger": {
    "runtime": "node"
  }
```

## Bun 

Set the field to `bun` to use the Bun runtime.

```json
  "dagger": {
    "runtime": "bun"
  }
```

:::warning

> The Bun runtime support is still experimental, Dagger cannot guarantee the perfect execution of the function until

Reviewer: ```suggestion
Bun runtime support is still experimental. Unexpected results may occur in some cases.
```

> Bun is fully compatible with Node. 

Reviewer: ```suggestion
```

:::note
For now, the TypeScript SDK only supports Node.js (stable) and Bun (experimental).
:::

To change the TypeScript SDK runtime, set the field `dagger.runtime` in the `package.json` file.

## Node  

Set the field to `node` to use the Node.js runtime. Node.js is also the default runtime used if this field is omitted.

```json
  "dagger": {
    "runtime": "node"
  }
```

## Bun 

Set the field to `bun` to use the Bun runtime.

```json
  "dagger": {
    "runtime": "bun"
  }
```

:::warning
Bun runtime support is still experimental. Unexpected results may occur in some cases.
:::
## Automatic Detection
When a runtime is not explicitly defined within the `package.json` file, Dagger automatically identifies the appropriate runtime by examining the project's lock files inside the project's `dagger` directory.

- If Dagger finds any of the following lock files : `package-lock.json`, `yarn.lock`, or `pnpm-lock.yaml`, it automatically selects `node` as the runtime.
- In the absence of the lock files mentioned above, if a `bun.lockb` file is present, Dagger will choose `bun` as the runtime.

- If no or only unknown lock files are present, `node` is chosen.

> This behavior however should be seen as a sensible fallback and not as an explicit configuration.

Reviewer 2: ```suggestion
:::warning
This behavior however should be seen as a sensible fallback and not as an explicit configuration.
:::
```

Maybe add this as a warning?

---
slug: /manuals/developer/typescript/356352/runtime
displayed_sidebar: "current"
toc_max_heading_level: 2
title: "Runtimes"
---

# Runtimes

> Multiple runtimes support TypeScript such as [Node.js](https://nodejs.org/en), [Deno](https://deno.com/) and [Bun](https://bun.sh/).

Reviewer: ```suggestion
TypeScript is supported by multiple runtimes: [Node.js](https://nodejs.org/en) (via tsx), [Deno](https://deno.com/) (natively) and [Bun](https://bun.sh/) (natively).
```
